### PR TITLE
Make PIN entry optional

### PR DESCRIPTION
**Pull request by nshutinicolas**
## Description

Currently MTN does not support PIN to be appended to the USSD string for mount equal or greater than 100 when purchasing airtime but Dialer still requires the PIN.

## The Fix
This PR:
- Makes the PIN entry optional for airtime purchase.
- Adds a new file `AppConstants` to use for all app constants
- Updates unit tests

### CODE REVIEW

File: Dialer/Dialer/View Models/MainViewModel.swift [hunk 1]
```diff
@@ -346,3 +345,17 @@ extension RecentDialCode {
...
+    var validAmount: Bool {
+        purchaseDetail.amount >= AppConstants.minAmount
+    }
+    
+    var validCode: Bool {
+        if let pinCode {
+            return String(pinCode).count == 5
+        }
+        return false
+    }
```
Comment (nshutinicolas) [inline Dialer/Dialer/View Models/MainViewModel.swift:360]: Moved this here to make it easy for testing in the near future

**Comment by cedricbahirwe** [comment]
Did you pull the latest changes before pushing ?

File: Dialer/DialerTests/Test Cases/ModelTests/PurchaseDetailModelTests.swift [hunk 2]
```diff
@@ -32,23 +32,23 @@ final class PurchaseDetailModelTests: XCTestCase {
...
         let purchase = PurchaseDetailModel(amount: 100)
 
-        let expectedPinnedCode = "*182*2*1*1*1*100*\(pin)#"
-        XCTAssertEqual(expectedPinnedCode, purchase.getDialCode(pin: "\(pin)"))
+        let expectedPinnedCode = "*182*2*1*1*1*100#"
+        XCTAssertEqual(expectedPinnedCode, purchase.getDialCode(pin: "\(pin)"), "For the range of 10 to 99, pin should not be appended to the string")
     }
     
-    func testPinnedPurchaseBigAmount() throws {
+    func testPinnedPurchase_amountInRangeForPinAppending() throws {
```
Comment (cedricbahirwe) [inline Dialer/DialerTests/Test Cases/ModelTests/PurchaseDetailModelTests.swift:43]: Adopt general Naming convention

File: Dialer/DialerTests/Test Cases/ModelTests/PurchaseDetailModelTests.swift [hunk 3]
```diff
@@ -32,23 +32,23 @@ final class PurchaseDetailModelTests: XCTestCase {
 
     }
 
-    func testPinnedPurchaseSmallAmount() throws {
+    func testPinnedPurchase_amountOutOfRangeForPinAppending() throws {
```
Comment (cedricbahirwe) [inline Dialer/DialerTests/Test Cases/ModelTests/PurchaseDetailModelTests.swift:35]: Adopt naming convetion here as well

File: Dialer/Dialer/Views/Main/PurchaseDetailView.swift [hunk 4]
```diff
@@ -152,11 +141,11 @@ struct PurchaseDetailView: View {
                             Label("Copy USSD code", systemImage: "doc.on.doc.fill")
                                 .frame(maxWidth: .infinity)
                                 .frame(height: 45)
-                                .background(Color.primary.opacity((!validCode || !validAmount) ? 0.5 : 1))
+                                .background(Color.primary.opacity((!data.validAmount) ? 0.5 : 1))
                                 .cornerRadius(8)
                                 .foregroundColor(Color(.systemBackground))
                         }
-                        .disabled(!validCode || !validAmount)
+                        .disabled(!data.validAmount)
```
Comment (cedricbahirwe) [inline Dialer/Dialer/Views/Main/PurchaseDetailView.swift:148]: `hasValidAmount` (this sounds like a boolean) vs `validAmount` (this seems to sound like a number)?!

File: Dialer/Dialer/View Models/MainViewModel.swift (hunk 1 above)
Comment (cedricbahirwe) [reply Dialer/Dialer/View Models/MainViewModel.swift:360]: Agreed, but I think the naming can be improve, check the comments below

File: Dialer/Dialer/Utilities/AppConstants.swift [hunk 5]
```diff
@@ -0,0 +1,14 @@
+//
+//  AppConstants.swift
+//  Dialer
+//
+//  Created by Nicolas Nshuti on 17/07/2023.
+//
+
+import Foundation
+
+struct AppConstants {
```
Comment (cedricbahirwe) [inline Dialer/Dialer/Utilities/AppConstants.swift:10]: Why a `struct`, why not an `enum` ??,

In this case of using a struct , someone might assume that the AppConstants can be initialized (`AppConstants()`), which is not correct 

If your data has multiple properties and represents objects with various attributes, use `struct`.
If your data has a limited set of related values or represents mutually exclusive options, use `enum`.

**Review by cedricbahirwe — CHANGES_REQUESTED** [review]
Welcomed PR 🙏 , Check minor comments

File: Dialer/DialerTests/Test Cases/ModelTests/PurchaseDetailModelTests.swift (hunk 2 above)
Comment (nshutinicolas) [reply Dialer/DialerTests/Test Cases/ModelTests/PurchaseDetailModelTests.swift:43]: Sorry about that. On it🫡

File: Dialer/Dialer/Views/Main/PurchaseDetailView.swift (hunk 4 above)
Comment (nshutinicolas) [reply Dialer/Dialer/Views/Main/PurchaseDetailView.swift:148]: Sure, it sounds better

File: Dialer/Dialer/Utilities/AppConstants.swift (hunk 5 above)
Comment (nshutinicolas) [reply Dialer/Dialer/Utilities/AppConstants.swift:10]: It is a struct as it holds different types.
Let me make the init private then

File: Dialer/Dialer/Utilities/AppConstants.swift [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    // Airtime USSD Dial constants
+    private init() { }
```
Comment (nshutinicolas) [inline Dialer/Dialer/Utilities/AppConstants.swift:12]: Made the init private to avoid initialization

**Review by cedricbahirwe — APPROVED** [review]
LGTM🔥

**Pull request merged**